File: backup/devfile.js
// Utility function for regex and enum validation
function validatePattern(value, pattern, field) {
  if (!new RegExp(pattern).test(value)) {
    throw new Error(`Invalid value for "${field}": "${value}" does not match pattern ${pattern}`);
  }
}
function validateEnum(value, enumArr, field) {
  if (!enumArr.includes(value)) {
    throw new Error(`Invalid value for "${field}": "${value}". Allowed: ${enumArr.join(", ")}`);
  }
}

class Devfile {
  constructor(data) {
    // Required
    if (!("schemaVersion" in data)) throw new Error("Missing required property: schemaVersion");
    this.schemaVersion = data.schemaVersion;
    validatePattern(this.schemaVersion, "^([2-9])\\.([0-9]+)\\.([0-9]+)(\\-[0-9a-z-]+(\\.[0-9a-z-]+)*)?(\\+[0-9A-Za-z-]+(\\.[0-9A-Za-z-]+)*)?$", "schemaVersion");

    // Optional
    this.attributes = data.attributes || {};
    this.commands = data.commands ? data.commands.map(c => new Command(c)) : [];
    this.components = data.components ? data.components.map(c => new Component(c)) : [];
    this.dependentProjects = data.dependentProjects ? data.dependentProjects.map(p => new Project(p)) : [];
    this.events = data.events ? new Events(data.events) : undefined;
    this.metadata = data.metadata ? new Metadata(data.metadata) : undefined;
    this.parent = data.parent ? new Parent(data.parent) : undefined;
    this.projects = data.projects ? data.projects.map(p => new Project(p)) : [];
    this.starterProjects = data.starterProjects ? data.starterProjects.map(p => new Project(p)) : [];
    this.variables = data.variables || {};
  }
}

class Command {
  constructor(data) {
    if (!("id" in data)) throw new Error("Missing required property: id in command");
    this.id = data.id;
    validatePattern(this.id, "^[a-z0-9]([-a-z0-9]*[a-z0-9])?$", "command.id");
    if (!("exec" in data || "apply" in data || "composite" in data))
      throw new Error("Command must have one of exec/apply/composite");
    if ("exec" in data) this.exec = new ExecCommand(data.exec);
    if ("apply" in data) this.apply = new ApplyCommand(data.apply);
    if ("composite" in data) this.composite = new CompositeCommand(data.composite);
    this.attributes = data.attributes || {};
  }
}
class ExecCommand {
  constructor(data) {
    ["commandLine", "component"].forEach(k => {
      if (!(k in data)) throw new Error(`Missing required "${k}" in exec command`);
    });
    this.commandLine = data.commandLine;
    this.component = data.component;
    this.env = data.env || [];
    this.group = data.group ? new CommandGroup(data.group) : undefined;
    this.hotReloadCapable = data.hotReloadCapable || false;
    this.label = data.label;
    this.workingDir = data.workingDir;
  }
}
class ApplyCommand {
  constructor(data) {
    if (!("component" in data)) throw new Error("Missing required 'component' in apply command");
    this.component = data.component;
    this.group = data.group ? new CommandGroup(data.group) : undefined;
    this.label = data.label;
  }
}
class CompositeCommand {
  constructor(data) {
    if ("commands" in data) this.commands = data.commands;
    this.group = data.group ? new CommandGroup(data.group) : undefined;
    this.label = data.label;
    this.parallel = data.parallel || false;
  }
}
class CommandGroup {
  constructor(data) {
    if (!("kind" in data)) throw new Error("Missing required 'kind' in command group");
    this.kind = data.kind;
    validateEnum(this.kind, ["build", "run", "test", "debug", "deploy"], "group.kind");
    this.isDefault = data.isDefault || false;
  }
}

class Component {
  constructor(data) {
    if (!("name" in data)) throw new Error("Missing 'name' in component");
    this.name = data.name;
    validatePattern(this.name, "^[a-z0-9]([-a-z0-9]*[a-z0-9])?$", "component.name");

    this.attributes = data.attributes || {};
    if ("container" in data) {
      this.container = new ContainerComponent(data.container);
    } else if ("kubernetes" in data) {
      this.kubernetes = new K8sComponent(data.kubernetes);
    } else if ("openshift" in data) {
      this.openshift = new OpenshiftComponent(data.openshift);
    } else if ("volume" in data) {
      this.volume = new VolumeComponent(data.volume);
    } else if ("image" in data) {
      this.image = new ImageComponent(data.image);
    } else {
      throw new Error("Component must have one of container/kubernetes/openshift/volume/image");
    }
  }
}
class ContainerComponent {
  constructor(data) {
    if (!("image" in data)) throw new Error("Missing required 'image' in container");
    this.image = data.image;
    this.annotation = data.annotation || {};
    this.args = data.args || [];
    this.command = data.command || [];
    this.cpuLimit = data.cpuLimit;
    this.cpuRequest = data.cpuRequest;
    this.dedicatedPod = data.dedicatedPod || false;
    this.endpoints = data.endpoints ? data.endpoints.map(e => new Endpoint(e)) : [];
    this.env = data.env || [];
    this.memoryLimit = data.memoryLimit;
    this.memoryRequest = data.memoryRequest;
    this.mountSources = data.mountSources !== undefined ? data.mountSources : true;
    this.sourceMapping = data.sourceMapping || "/projects";
    this.volumeMounts = data.volumeMounts ? data.volumeMounts.map(vm => new VolumeMount(vm)) : [];
  }
}
class Endpoint {
  constructor(data) {
    ["name", "targetPort"].forEach(k => { if (!(k in data)) throw new Error(`Missing required "${k}" in endpoint`); });
    this.name = data.name;
    validatePattern(this.name, "^[a-z0-9]([-a-z0-9]*[a-z0-9])?$", "endpoint.name");
    this.targetPort = data.targetPort;
    this.protocol = data.protocol || "http";
    validateEnum(this.protocol, ["http", "https", "ws", "wss", "tcp", "udp"], "endpoint.protocol");
    this.exposure = data.exposure || "public";
    validateEnum(this.exposure, ["public", "internal", "none"], "endpoint.exposure");
    this.secure = data.secure || false;
    this.path = data.path;
    this.annotation = data.annotation || {};
    this.attributes = data.attributes || {};
  }
}
class VolumeMount {
  constructor(data) {
    if (!("name" in data)) throw new Error("Missing 'name' in volume mount");
    this.name = data.name;
    validatePattern(this.name, "^[a-z0-9]([-a-z0-9]*[a-z0-9])?$", "volumeMount.name");
    this.path = data.path || `/${data.name}`;
  }
}
class K8sComponent {
  constructor(data) {
    if (!("uri" in data || "inlined" in data)) throw new Error("Kubernetes component requires 'uri' or 'inlined'");
    this.uri = data.uri;
    this.inlined = data.inlined;
    this.deployByDefault = data.deployByDefault || false;
    this.endpoints = data.endpoints ? data.endpoints.map(e => new Endpoint(e)) : [];
  }
}
class OpenshiftComponent {
  constructor(data) {
    if (!("uri" in data || "inlined" in data)) throw new Error("Openshift component requires 'uri' or 'inlined'");
    this.uri = data.uri;
    this.inlined = data.inlined;
    this.deployByDefault = data.deployByDefault || false;
    this.endpoints = data.endpoints ? data.endpoints.map(e => new Endpoint(e)) : [];
  }
}
class VolumeComponent {
  constructor(data) {
    this.ephemeral = data.ephemeral || false;
    this.size = data.size;
  }
}
class ImageComponent {
  constructor(data) {
    if (!("imageName" in data)) throw new Error("Missing 'imageName' in image component");
    this.imageName = data.imageName;
    if ("dockerfile" in data) this.dockerfile = new Dockerfile(data.dockerfile);
    this.autoBuild = data.autoBuild || false;
  }
}
class Dockerfile {
  constructor(data) {
    if (!("uri" in data || "devfileRegistry" in data || "git" in data)) {
      throw new Error("Dockerfile requires uri, devfileRegistry or git");
    }
    this.args = data.args || [];
    this.buildContext = data.buildContext;
    this.rootRequired = data.rootRequired || false;
    this.uri = data.uri;
    this.devfileRegistry = data.devfileRegistry;
    this.git = data.git;
  }
}

// Project and StarterProject are similar
class Project {
  constructor(data) {
    if (!("name" in data)) throw new Error("Missing 'name' in project");
    this.name = data.name;
    validatePattern(this.name, "^[a-z0-9]([-a-z0-9]*[a-z0-9])?$", "project.name");
    if (!("git" in data || "zip" in data))
      throw new Error("Project must have either 'git' or 'zip' source");
    if ("git" in data) this.git = new GitSource(data.git);
    if ("zip" in data) this.zip = new ZipSource(data.zip);
    this.clonePath = data.clonePath || this.name;
    this.attributes = data.attributes || {};
  }
}
class GitSource {
  constructor(data) {
    if (!("remotes" in data)) throw new Error("Missing 'remotes' in git source");
    this.remotes = data.remotes;
    this.checkoutFrom = data.checkoutFrom;
  }
}
class ZipSource {
  constructor(data) {
    if (!("location" in data)) throw new Error("Missing 'location' in zip source");
    this.location = data.location;
  }
}

class Events {
  constructor(data) {
    this.preStart = data.preStart || [];
    this.postStart = data.postStart || [];
    this.preStop = data.preStop || [];
    this.postStop = data.postStop || [];
  }
}

class Metadata {
  constructor(data) {
    this.architectures = data.architectures || [];
    if (this.architectures.length) {
      this.architectures.forEach(a =>
        validateEnum(a, ["amd64", "arm64", "ppc64le", "s390x"], "metadata.architectures"));
    }
    this.attributes = data.attributes || {};
    this.description = data.description;
    this.displayName = data.displayName;
    this.globalMemoryLimit = data.globalMemoryLimit;
    this.icon = data.icon;
    this.language = data.language;
    this.name = data.name;
    this.projectType = data.projectType;
    this.provider = data.provider;
    this.supportUrl = data.supportUrl;
    this.tags = data.tags || [];
    this.version = data.version;
    if (this.version)
      validatePattern(this.version, "^([0-9]+)\\.([0-9]+)\\.([0-9]+)(\\-[0-9a-z-]+(\\.[0-9a-z-]+)*)?(\\+[0-9A-Za-z-]+(\\.[0-9A-Za-z-]+)*)?$", "metadata.version");
    this.website = data.website;
  }
}

class Parent {
  constructor(data) {
    if (!("uri" in data || "id" in data || "kubernetes" in data))
      throw new Error("Parent must have 'uri', 'id' or 'kubernetes'");
    this.uri = data.uri;
    this.id = data.id;
    this.registryUrl = data.registryUrl;
    this.version = data.version;
    this.attributes = data.attributes || {};
    this.commands = data.commands ? data.commands.map(c => new Command(c)) : [];
    this.components = data.components ? data.components.map(c => new Component(c)) : [];
    this.dependentProjects = data.dependentProjects ? data.dependentProjects.map(p => new Project(p)) : [];
    this.projects = data.projects ? data.projects.map(p => new Project(p)) : [];
    this.starterProjects = data.starterProjects ? data.starterProjects.map(p => new Project(p)) : [];
    this.variables = data.variables || {};
    this.kubernetes = data.kubernetes;
  }
}

// Usage Example:
// const devfile = new Devfile(devfileJson);
// devfile will be fully validated and structured

export {
  Devfile, Command, Component, Project, Events, Metadata, Parent,
};

File: src/app/service/devfile.py
from __future__ import annotations

import copy
from typing import Any


# ── Fragment merge ────────────────────────────────────────────────────────────


def _merge_fragments(
    base: dict[str, Any],
    fragments: list[dict[str, Any]],
) -> tuple[dict[str, Any], list[str]]:
    """
    Deep-merge a list of partial devfile fragment dicts into ``base``.

    Each fragment dict has the shape:
      { id, name, description, icon, devfile: { components, commands, events,
        variables, attributes, projects, starterProjects } }

    Merge rules (mirrors mergeDevfileFragments in devfile.js):
      schemaVersion, metadata  — base wins; fragments cannot override
      variables, attributes    — shallow merge; base wins on key conflicts
      components               — array concat; duplicate names → error
      projects, starterProjects — array concat; duplicate names → error
      commands                 — array concat; duplicate IDs → error
      events.*                 — array union (deduplicated)

    Returns (merged_devfile, error_list).
    """
    merged = copy.deepcopy(base)
    errors: list[str] = []

    # Seed uniqueness sets from the base
    component_names: set[str] = {
        c.get("name", "") for c in merged.get("components") or []
    }
    project_names: set[str] = {p.get("name", "") for p in merged.get("projects") or []}
    starter_names: set[str] = {
        s.get("name", "") for s in merged.get("starterProjects") or []
    }
    command_ids: set[str] = {c.get("id", "") for c in merged.get("commands") or []}

    for frag in fragments:
        label = frag.get("name") or frag.get("id") or "(unnamed)"
        d: dict[str, Any] = frag.get("devfile") or {}

        # variables — base wins on conflict
        for k, v in (d.get("variables") or {}).items():
            if k not in (merged.get("variables") or {}):
                merged.setdefault("variables", {})[k] = v

        # attributes — base wins on conflict
        for k, v in (d.get("attributes") or {}).items():
            if k not in (merged.get("attributes") or {}):
                merged.setdefault("attributes", {})[k] = v

        # components — concat, error on duplicate name
        for comp in d.get("components") or []:
            name = comp.get("name", "")
            if name in component_names:
                errors.append(
                    f'Fragment "{label}": component name "{name}" already exists.'
                )
            else:
                merged.setdefault("components", []).append(comp)
                component_names.add(name)

        # projects — concat, error on duplicate name
        for proj in d.get("projects") or []:
            name = proj.get("name", "")
            if name in project_names:
                errors.append(
                    f'Fragment "{label}": project name "{name}" already exists.'
                )
            else:
                merged.setdefault("projects", []).append(proj)
                project_names.add(name)

        # starterProjects — concat, error on duplicate name
        for sp in d.get("starterProjects") or []:
            name = sp.get("name", "")
            if name in starter_names:
                errors.append(
                    f'Fragment "{label}": starterProject name "{name}" already exists.'
                )
            else:
                merged.setdefault("starterProjects", []).append(sp)
                starter_names.add(name)

        # commands — concat, error on duplicate ID
        for cmd in d.get("commands") or []:
            cmd_id = cmd.get("id", "")
            if cmd_id in command_ids:
                errors.append(
                    f'Fragment "{label}": command ID "{cmd_id}" already exists.'
                )
            else:
                merged.setdefault("commands", []).append(cmd)
                command_ids.add(cmd_id)

        # events — union per lifecycle key
        frag_events = d.get("events") or {}
        for key in ("preStart", "postStart", "preStop", "postStop"):
            for cmd_id in frag_events.get(key) or []:
                merged.setdefault("events", {}).setdefault(key, [])
                if cmd_id not in merged["events"][key]:
                    merged["events"][key].append(cmd_id)

    return merged, errors


def _generate_devfile_data(state: dict[str, Any]) -> dict[str, Any]:
    """
    Build a final devfile dict from the Alpine store state shape.

    Builds a base devfile from the store's structural fields, then
    deep-merges each fragment in ``state["fragments"]`` in order.
    """
    components_raw: list[dict] = state.get("components") or []
    first_comp = components_raw[0] if components_raw else {}
    container_name: str = first_comp.get("name") or "dev"
    ct: dict[str, Any] = first_comp.get("container") or {}

    env_vars = generate_env_vars(state)
    vol_components, vol_mounts = generate_volume_components_and_mounts(state)
    endpoints = generate_endpoints(state)
    container_component = generate_container_component(
        container_name, ct, env_vars, endpoints, vol_mounts
    )

    components = [container_component, *vol_components]
    projects = generate_projects(state)
    starter_projects = generate_starter_projects(state)
    commands = generate_commands(state)
    events = generate_events(state)

    # ── Top-level optional fields ──────────────────────────────────────────────
    meta_raw: dict = state.get("metadata") or {}
    meta_out: dict[str, Any] = {}
    for field in (
        "name",
        "displayName",
        "description",
        "version",
        "language",
        "projectType",
        "provider",
        "supportUrl",
        "website",
        "icon",
    ):
        if meta_raw.get(field):
            meta_out[field] = meta_raw[field]
    if meta_raw.get("tags"):
        meta_out["tags"] = meta_raw["tags"]
    if meta_raw.get("architectures"):
        meta_out["architectures"] = meta_raw["architectures"]
    if meta_raw.get("attributes"):
        meta_out["attributes"] = meta_raw["attributes"]

    variables: dict = state.get("variables") or {}
    top_attributes: dict = state.get("attributes") or {}

    # ── Assemble base devfile ──────────────────────────────────────────────────
    base: dict[str, Any] = {"schemaVersion": "2.3.0"}
    if meta_out:
        base["metadata"] = meta_out
    if variables:
        base["variables"] = variables
    if top_attributes:
        base["attributes"] = top_attributes
    base["components"] = components
    if projects:
        base["projects"] = projects
    if starter_projects:
        base["starterProjects"] = starter_projects
    if commands:
        base["commands"] = commands
    if events:
        base["events"] = events

    # ── Merge fragments ────────────────────────────────────────────────────────
    fragments: list[dict] = state.get("fragments") or []
    merged, errors = _merge_fragments(base, fragments)
    if errors:
        raise ValueError("\n".join(errors))

    return merged


def generate_env_vars(state: dict[str, Any]) -> list[dict]:
    env_vars: list[dict] = []
    for e in state.get("customEnv") or []:
        if e.get("name"):
            env_vars.append({"name": e["name"], "value": e.get("value", "")})
    return env_vars


def generate_volume_components_and_mounts(
    state: dict[str, Any],
) -> tuple[list[dict], list[dict]]:
    vol_components: list[dict] = []
    vol_mounts: list[dict] = []
    for v in state.get("volumes") or []:
        name = v.get("name", "").strip()
        if not name:
            continue
        vol_def: dict = {}
        if v.get("size"):
            vol_def["size"] = v["size"]
        if v.get("ephemeral"):
            vol_def["ephemeral"] = True
        vol_components.append({"name": name, "volume": vol_def})
        vol_mounts.append({"name": name, "path": v.get("mountPath") or f"/{name}"})
    return vol_components, vol_mounts


def generate_endpoints(state: dict[str, Any]) -> list[dict]:
    out = []
    for ep in state.get("endpoints") or []:
        if not (ep.get("name") and ep.get("targetPort")):
            continue
        entry: dict[str, Any] = {
            "name": ep["name"],
            "targetPort": ep["targetPort"],
            "protocol": ep.get("protocol", "http"),
            "exposure": ep.get("exposure", "public"),
        }
        if ep.get("secure"):
            entry["secure"] = True
        if ep.get("path"):
            entry["path"] = ep["path"]
        if ep.get("annotation"):
            entry["annotation"] = ep["annotation"]
        out.append(entry)
    return out


def generate_container_component(
    name: str,
    ct: dict[str, Any],
    env_vars: list[dict],
    endpoints: list[dict],
    vol_mounts: list[dict],
) -> dict[str, Any]:
    container: dict[str, Any] = {
        "image": ct.get("image") or "quay.io/devfile/universal-developer-image:latest",
        "mountSources": ct.get("mountSources", True) is not False,
    }
    if ct.get("cpuRequest"):
        container["cpuRequest"] = ct["cpuRequest"]
    if ct.get("cpuLimit"):
        container["cpuLimit"] = ct["cpuLimit"]
    if ct.get("memoryRequest"):
        container["memoryRequest"] = ct["memoryRequest"]
    if ct.get("memoryLimit"):
        container["memoryLimit"] = ct["memoryLimit"]
    if ct.get("sourceMapping"):
        container["sourceMapping"] = ct["sourceMapping"]
    if ct.get("dedicatedPod"):
        container["dedicatedPod"] = True
    if ct.get("args"):
        container["args"] = ct["args"]
    if ct.get("command"):
        container["command"] = ct["command"]
    if ct.get("annotation"):
        container["annotation"] = ct["annotation"]
    if env_vars:
        container["env"] = env_vars
    if endpoints:
        container["endpoints"] = endpoints
    if vol_mounts:
        container["volumeMounts"] = vol_mounts

    return {"name": name or "dev", "container": container}


def generate_projects(state: dict[str, Any]) -> list[dict]:
    projects = []
    for p in state.get("projects") or []:
        name = p.get("name", "")
        git = p.get("git") or {}
        remotes = git.get("remotes") or {}
        origin = remotes.get("origin", "")
        if not origin:
            origin = p.get("remote", "") or next(iter(remotes.values()), "")
        if not name or not origin:
            continue
        revision = (git.get("checkoutFrom") or {}).get("revision", "main")
        checkout_from: dict[str, str] = {}
        if revision and revision != "main":
            checkout_from["revision"] = revision
        git_out: dict[str, Any] = {"remotes": {"origin": origin}}
        if checkout_from:
            git_out["checkoutFrom"] = checkout_from
        proj: dict[str, Any] = {"name": name, "git": git_out}
        clone_path = p.get("clonePath", "")
        if clone_path and clone_path != name:
            proj["clonePath"] = clone_path
        if p.get("attributes"):
            proj["attributes"] = p["attributes"]
        projects.append(proj)
    return projects


def generate_starter_projects(state: dict[str, Any]) -> list[dict]:
    """
    Build the starterProjects list from the new state shape:
      starterProjects: [ { name, git: { remotes: { origin }, checkoutFrom: { revision } } } ]
    """
    starter_projects = []
    for s in state.get("starterProjects") or []:
        name = s.get("name", "")
        git = s.get("git") or {}
        remotes = git.get("remotes") or {}
        origin = remotes.get("origin", "")

        if not origin:
            origin = s.get("remote", "") or next(iter(remotes.values()), "")

        if not name or not origin:
            continue

        revision = (git.get("checkoutFrom") or {}).get("revision", "main")
        checkout_from: dict[str, str] = {}
        if revision and revision != "main":
            checkout_from["revision"] = revision

        git_out: dict[str, Any] = {"remotes": {"origin": origin}}
        if checkout_from:
            git_out["checkoutFrom"] = checkout_from

        starter_projects.append({"name": name, "git": git_out})

    return starter_projects


def generate_commands(state: dict[str, Any]) -> list[dict]:
    commands = []
    for c in state.get("commands") or []:
        if not c.get("id"):
            continue
        entry: dict[str, Any] = {"id": c["id"]}
        if c.get("attributes"):
            entry["attributes"] = c["attributes"]
        cmd_type = c.get("type")
        if cmd_type == "exec" and c.get("exec"):
            generate_exec_command(c, entry)
        elif cmd_type == "composite" and c.get("composite"):
            generate_composite_command(c, entry)
        elif cmd_type == "apply" and c.get("apply"):
            generate_apply_command(c, entry)
        else:
            continue
        commands.append(entry)

    return commands


def generate_exec_command(c: dict[str, Any], entry: dict[str, Any]) -> None:
    ex = c["exec"]
    exec_obj: dict[str, Any] = {
        "commandLine": ex.get("commandLine", ""),
        "component": ex.get("component", ""),
    }
    if ex.get("workingDir"):
        exec_obj["workingDir"] = ex["workingDir"]
    if ex.get("label"):
        exec_obj["label"] = ex["label"]
    if ex.get("hotReloadCapable"):
        exec_obj["hotReloadCapable"] = True
    if ex.get("env"):
        exec_obj["env"] = ex["env"]
    if ex.get("group") and ex["group"].get("kind"):
        exec_obj["group"] = {
            "kind": ex["group"]["kind"],
            "isDefault": bool(ex["group"].get("isDefault")),
        }
    entry["exec"] = exec_obj


def generate_composite_command(c: dict[str, Any], entry: dict[str, Any]) -> None:
    cp = c["composite"]
    comp_obj: dict[str, Any] = {}
    if cp.get("commands"):
        comp_obj["commands"] = cp["commands"]
    if cp.get("parallel"):
        comp_obj["parallel"] = True
    if cp.get("label"):
        comp_obj["label"] = cp["label"]
    if cp.get("group") and cp["group"].get("kind"):
        comp_obj["group"] = {
            "kind": cp["group"]["kind"],
            "isDefault": bool(cp["group"].get("isDefault")),
        }
    entry["composite"] = comp_obj


def generate_apply_command(c: dict[str, Any], entry: dict[str, Any]) -> None:
    ap = c["apply"]
    apply_obj: dict[str, Any] = {"component": ap.get("component", "")}
    if ap.get("label"):
        apply_obj["label"] = ap["label"]
    if ap.get("group") and ap["group"].get("kind"):
        apply_obj["group"] = {
            "kind": ap["group"]["kind"],
            "isDefault": bool(ap["group"].get("isDefault")),
        }
    entry["apply"] = apply_obj


def generate_events(state: dict[str, Any]) -> dict[str, list[str]]:
    raw_events = state.get("events") or {}
    events: dict[str, list[str]] = {}
    for key in ("preStart", "postStart", "preStop", "postStop"):
        ids = [i for i in (raw_events.get(key) or []) if i]
        if ids:
            events[key] = ids
    return events
